doc/decisions/lowlevel_library.md

```diff
@@ -0,0 +1,61 @@
+# Operations Library
+
+## Problem
+
+There are many functions that can be useful, but are not required to form the minimal API.
+Since we want to keep the API of `libelektra-core` minimal, we cannot add these functions.
+
+If a function is generally useful, it should be added to [`libelektra-operations`](operations_library.md).
+However, some functions are only useful as a C API and shouldn't be part of a language binding.
+
+## Constraints
+
+- `libelektra-core` should have a minimal (but still useful) API.
+- `libelektra-operations` should contain generally useful, but not minimal API, that can be used via language bindings
+
+## Assumptions
+
+## Considered Alternatives
+
+- Put everything into `libelektra-operations`
+
+## Decision
+
+Create a new `libelektra-lowlevel` library (as a counterpart to `libelektra-highlevel`).
+This library provides functions which are:
+
+- not minimal
+- useful to C code
+- but not useful or not usable outside of C code
+
+The implementations should still be as efficient as possible and will therefore access private API directly.
+
+`libelektra-lowlevel` depends on `libelektra-operations` and therefore on `libelektra-core`.
+
+As an example for what should be part of this library take a look at [`keyGetNextPart`](iterating_name_parts.md.md).
+It can be useful to C code (especially for beginners), but it clearly shouldn't be used in any language that has its own iterator type.
+
+Other examples include `elektraKeyCreate` and `elektraKsCreate`.
+Their API allows the creation of an entire KeySet within a single expression.
+However, the API was specially designed for use in C and is very hard or impossible to use from other languages.
+This makes them the ideal candidates for `libelektra-lowlevel`.
+
+## Rationale
+
+The separation between `libelektra-operations` and `libelektra-lowlevel` has a few advantages:
+
+- It is clear what the target audience for the API is (everyone vs. C-only)
```

Comment:
I fully agree that (at least) two different libraries based on this distinction must exist.

Open questions are:

- is it only 2?
- what is the best name? ("lowlevel" actually suggests that a binding for it might be useful, as "highlevel" also is open to bindings)

Reply:
> is it only 2?

For now there is not enough proposed functionality to justify more. We can always break apart e.g. `libelektra-operations`  in future. AFAIK this could be backwards compatible, as long as we continue to ship a `libelektra-operations` that simply links to the libraries that actually implement the functions.

However, if the library isn't big enough to justify a split before 1.0, I don't think it will grow so fast that we need to split it, before we have some other reason to go to 2.0. I say this, because there would still be `libelektra-ease` (possibly with a different) for more rarely using things (e.g. topological sort) and others like `libelektra-opts` for specialized things.

> as "highlevel" also is open to bindings

The current `libelektra-highlevel` is really not all that open to bindings. It would make a lot more sense to implement high-level abstractions on top of low-level bindings of other languages. The concepts behind the library (shorter key names, check on open, no fail in get, ...) and maybe even the code-generation could be applied to other languages. But the actual API of `libelektra-highlevel` is really only suitable for C or a similarly basic language. For example, in an object oriented language it would make much more sense to utilise classes. Some cases could even produce an API (with code generation) that utilises the type system, e.g. if it compiles you won't get "key not found" errors.

Reply:
> For now there is not enough proposed functionality to justify more. We can always break apart e.g. libelektra-operations in future. AFAIK this could be backwards compatible, as long as we continue to ship a libelektra-operations that simply links to the libraries that actually implement the functions.

It is much better to get it right in the first place. Which symbols are in which library is part of the API.

> However, if the library isn't big enough to justify a split before 1.0, I don't think it will grow so fast that we need to split it, before we have some other reason to go to 2.0.

We cannot even do such a big change in 2.0 as it basically breaks every application using these libs. It is also very annoying (speak: deal breaker) for developers to constantly relearn in which header/libs they find which kind of functionality.

> The current libelektra-highlevel is really not all that open to bindings.

Ok, then call it "lowlevel" makes sense. We need to be careful to only have functionality there that is really only useful for C. Maybe we should clarify the name even more by calling them libelektra-lowlevel-c and libelektra-highlevel-c?

Reply:
> Which symbols are in which library is part of the API.

Let's say 1.0 has `libelektra-operations.so` which contains the implementations for `foo` and `bar` and comes with the header `elektra/operations.h`

```c
int foo (int arg);
char * bar(const char * x, int y);
```

Now let's say in 2.0 we decide to split this library. We split `libelektra-operations.so` into `libelektra-foo.so` and `libelektra-bar.so`. `libelektra-foo.so` contains the implementation for `foo` and `libelektra-bar.so` the implementation for `bar`. We also add new headers `elektra/foo.h`

```c
int foo (int arg);
```

and `elektra/bar.h`

```c
char * bar(const char * x, int y);
```

If we only do this, it would be major breaking change and everything that depends on these functions would have to be rebuilt. IMHO this would be fine, since that is basically what breaking change means. However, we could continue to ship  `libelektra-operations.so` and `elektra/operations.h` to avoid the breaking change.

`libelektra-operations.so` would simply have dependencies on `libelektra-foo.so` and `libelektra-bar.so`. The system linker would then load these at startup. Similarly, `elektra/operations.h` would now look like this

```c
#include "foo.h"
#include "bar.h"
```

to avoid source-breaking changes.

> Maybe we should clarify the name even more by calling them libelektra-lowlevel-c and libelektra-highlevel-c?

Yes, that is probably a good idea.